datafusion/optimizer/src/eliminate_outer_join.rs

```diff
@@ -443,4 +361,214 @@ mod tests {
+        \n    TableScan: t1\
+        \n    TableScan: t2";
+        assert_optimized_plan_equal(plan, expected)
+    }
+
+    #[derive(Debug)]
+    struct DoNothingUdf {
+        signature: Signature,
+    }
+
+    impl DoNothingUdf {
+        pub fn new() -> Self {
+            Self {
+                signature: Signature::any(1, Volatility::Immutable),
+            }
+        }
+    }
+
+    impl ScalarUDFImpl for DoNothingUdf {
+        fn as_any(&self) -> &dyn std::any::Any {
+            self
+        }
+
+        fn name(&self) -> &str {
+            "do_nothing"
+        }
+
+        fn signature(&self) -> &Signature {
+            &self.signature
+        }
+
+        fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+            Ok(arg_types[0].clone())
+        }
+
+        fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+            Ok(args[0].clone())
+        }
+    }
+
+    #[test]
+    fn eliminate_right_with_udf() -> Result<()> {
+        let t1 = test_table_scan_with_name("t1")?;
+        let t2 = test_table_scan_with_name("t2")?;
+        let fun = Arc::new(ScalarUDF::new_from_impl(DoNothingUdf::new()));
+
+        let plan = LogicalPlanBuilder::from(t1)
+            .join(
+                t2,
+                JoinType::Right,
+                (vec![Column::from_name("a")], vec![Column::from_name("a")]),
+                None,
+            )?
+            .filter(
+                Expr::ScalarFunction(ScalarFunction::new_udf(fun, vec![col("t1.b")]))
+                    .gt(lit(10u32)),
+            )?
+            .build()?;
+
+        let expected = "\
+        Filter: do_nothing(t1.b) > UInt32(10)\
+        \n  Inner Join: t1.a = t2.a\
+        \n    TableScan: t1\
+        \n    TableScan: t2";
+        assert_optimized_plan_equal(plan, expected)
+    }
+
+    #[derive(Debug)]
+    struct AlwaysNullUdf {
+        signature: Signature,
+    }
+
+    impl AlwaysNullUdf {
+        pub fn new() -> Self {
+            Self {
+                signature: Signature::any(1, Volatility::Immutable),
+            }
+        }
+    }
+
+    impl ScalarUDFImpl for AlwaysNullUdf {
+        fn as_any(&self) -> &dyn std::any::Any {
+            self
+        }
+
+        fn name(&self) -> &str {
+            "always_null"
+        }
+
+        fn signature(&self) -> &Signature {
+            &self.signature
+        }
+
+        fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+            Ok(DataType::Null)
+        }
+
+        fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+            Ok(match &args[0] {
+                ColumnarValue::Array(array) => {
+                    ColumnarValue::create_null_array(array.len())
+                }
+                ColumnarValue::Scalar(_) => ColumnarValue::Scalar(ScalarValue::Null),
+            })
+        }
+    }
+
+    #[test]
+    fn eliminate_right_with_null_udf() -> Result<()> {
```

Comment:
The test name makes it should to be that we expect it to eliminate the join, but we do not?

Reply:
Since `always_null(t1.b) IS NULL` is not a restrict null predicate, it cannot eliminate the join. This is probably repeated with `eliminate_left_with_null`

Reply:
My suggestion would be to name the test `not_eliminate` since we do not expect the test case to eliminate the outer join.